spec/controllers/concerns/course/lesson_plan/personalization_concern_spec.rb

```diff
@@ -47,20 +52,54 @@ class self::DummyController < ApplicationController
 
     context 'when course user is on the stragglers algorithm' do
       let!(:course_user) { create(:course_user, course: course, timeline_algorithm: 'stragglers') }
-      let!(:submission1) do
-        create(:course_assessment_submission, assessment: assessment1, creator: course_user.user).tap(&:finalise!)
+
+      def submit_assessment(assessment)
+        create(:course_assessment_submission, assessment: assessment, creator: course_user.user).
+          tap(&:finalise!)
       end
 
       it 'creates personal times' do
+        submit_assessment(assessment)
         dummy_controller.send(:update_personalized_timeline_for, course_user)
         expect(course_user.personal_times.count).to eq(course.assessments.count - 1)
       end
+
+      it 'shifts the end_at of non-open items forward' do
+        submit_assessment(overdue_assessment)
+        dummy_controller.send(:update_personalized_timeline_for, course_user)
+        original_end_at = yet_to_open_assessment.lesson_plan_item.personal_time_for(course_user).end_at
+
+        submit_assessment(assessment)
+        dummy_controller.send(:update_personalized_timeline_for, course_user)
+        new_end_at = yet_to_open_assessment.lesson_plan_item.personal_time_for(course_user).end_at
+        expect(new_end_at).to be < original_end_at
+      end
+
+      it 'does not shift the end_at of already open items forward' do
+        submit_assessment(overdue_assessment)
+        dummy_controller.send(:update_personalized_timeline_for, course_user)
+        original_end_at = already_open_assessment.lesson_plan_item.personal_time_for(course_user).end_at
+
+        submit_assessment(assessment)
+        dummy_controller.send(:update_personalized_timeline_for, course_user)
+        new_end_at = already_open_assessment.lesson_plan_item.personal_time_for(course_user).end_at
+
+        expect(new_end_at).to eq(original_end_at)
+      end
+
+      it 'rounds off to 2359' do
+        submit_assessment(overdue_assessment)
+        dummy_controller.send(:update_personalized_timeline_for, course_user)
+        end_at = assessment.lesson_plan_item.personal_time_for(course_user).end_at
+        course_tz = course.time_zone
+        expect(end_at.in_time_zone(course_tz).strftime('%H:%M')).to eq('23:59')
```

Comment:
Do we want to check the date as well?

Reply:
Actually may not be ideal. As mentioned in the PR description, the exact amount shifted is quite hard to "predict", in that we might not want to test it because it will likely flake once we change anything with the algorithm. We instead just want to ensure that `23:59` is the time we get, which is a property that is desirable and won't flake as easily.

Reply:
Thanks for clarifying and all good to go!